src/interactivity/base/HostSignalInputThread.cpp

```diff
@@ -0,0 +1,240 @@
+//      (in and screen buffers) haven't yet been initialized.
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HostSignalInputThread::ConnectConsole() noexcept
+{
+    _consoleConnected = true;
+}
+
+// Method Description:
+// - Attempts to retrieve a given type T off of the internal
+//   pipe channel and return it.
+// Return Value:
+// - A structure filled with the specified data off the byte stream
+// - EXCEPTIONS may be thrown if the packet size mismatches
+//     or if we fail to read for another reason.
+template<typename T>
+T HostSignalInputThread::_ReceiveTypedPacket()
+{
+    T msg = { 0 };
+    THROW_HR_IF(E_ABORT, !_GetData(&msg, sizeof(msg)));
+
+    // If the message size was stated to be larger, we
+    // want to seek forward to the next message code.
+    // If it's equal, we'll seek forward by 0 and
+    // do nothing.
+    if (msg.sizeInBytes >= sizeof(msg))
+    {
+        _AdvanceReader(msg.sizeInBytes - sizeof(msg));
+    }
+    // If the message is smaller than what we expected
+    // then it was malformed and we need to throw.
+    else
+    {
+        THROW_HR(E_ILLEGAL_METHOD_CALL);
+    }
+
+    return msg;
+}
+
+// Method Description:
+// - The ThreadProc for the Host Signal Input Thread.
+// Return Value:
+// - S_OK if the thread runs to completion.
+// - Otherwise it may cause an application termination and never return.
+[[nodiscard]] HRESULT HostSignalInputThread::_InputThread()
+{
+    HostSignals signalId;
+    while (_GetData(&signalId, sizeof(signalId)))
+    {
+        switch (signalId)
+        {
+        case HostSignals::NotifyApp:
+        {
+            auto msg = _ReceiveTypedPacket<HostSignalNotifyAppData>();
+
+            LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->NotifyConsoleApplication(msg.processId));
+
+            break;
+        }
+        case HostSignals::SetForeground:
+        {
+            auto msg = _ReceiveTypedPacket<HostSignalSetForegroundData>();
+
+            LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->SetForeground(ULongToHandle(msg.processId), msg.isForeground));
+
+            break;
+        }
+        case HostSignals::EndTask:
+        {
+            auto msg = _ReceiveTypedPacket<HostSignalEndTaskData>();
+
+            LOG_IF_NTSTATUS_FAILED(ServiceLocator::LocateConsoleControl()->EndTask(ULongToHandle(msg.processId), msg.eventType, msg.ctrlFlags));
+
+            break;
+        }
+        default:
+        {
+            THROW_HR(E_UNEXPECTED);
+            break;
+        }
+        }
+    }
+    return S_OK;
+}
+
+// Method Description:
+// - Skips the file stream forward by the specified number of bytes.
+// Arguments:
+// - cbBytes - Count of bytes to skip forward
+// Return Value:
+// - True if we could skip forward successfully. False otherwise.
+bool HostSignalInputThread::_AdvanceReader(const DWORD cbBytes)
+{
+    std::array<BYTE, 8> buffer;

```

Comment:
It can't hurt making this larger. A cache line is at least 64 Bytes nowadays, which can be the minimum for any buffer like that. A weak min-spec CPU usually has about 2-4MB L3 cache at least.
If you make this 256-1024 bytes it's just 5-17 cache lines at most - of which only those are swapped into CPU cache, that are actually touched by this function. And finally: Stack memory is cheap after all...

Reply:
256 it is.